### PR TITLE
Fix replacement of dot with variable in script fragments containing chained function calls

### PR DESCRIPTION
**Pull request by SchmartMaker**
This PR adjusts the regular expression that is used in the code that supports "dot syntax".

The current regular expression does not sufficiently account for (string) functions applied to a variable; script fragments such as ` .Title.TrimEnd().EndsWith('Word')` will fail execution, because the third dot in this script is also seen as a shortcut of the current context variable, which result in a script compiler exception.

The PR:

- Adjusts the regex for variable replacement slightly to recognize functions.
- Adds a unit test to assure expressions using chained functions work.
- Adds an end-to-end unit test demonstrating actual document generation.

All unit tests have been run and complete successfully. It is therefore assumed the PR has no unintentional side-effects (also not with double-dot syntax `..`).

Also see: https://github.com/Amberg/DocxTemplater/issues/66

### CODE REVIEW

**Comment by Amberg** [comment]
Hi
Thank you for contributing. A "end to end" test would be nice. Otherwise the "PatternMatcher" is not tested
Like "https://github.com/Amberg/DocxTemplater/blob/c4949ffb7781a0a433d5ae6a821328c2ab2e391d/DocxTemplater.Test/DocxTemplateTest.cs#L520"

**Comment by SchmartMaker** [comment]
I think I've come across an edge case with `.Func()`. I still have to confirm the case and would like to potentially continue further investigation and development in a separate PR. I would appreciate some disscussion at a later time if you had forseen the use of script syntax like `.Func()` or `..Func()`, where there's no property name after the dots.

Also, I'm considering expanding the regex to support a string array accessor, so the individual characters of a string can be accessed with the syntax `.[i]` or `.Property[i]`.

File: DocxTemplater/ScriptCompiler.cs [hunk 1]
```diff
@@ -11,14 +11,17 @@ internal class ScriptCompiler : IScriptCompiler
...
+                                                                    (?:                          # Non-capturing group for all possible prefixes:
+                                                                        ^                        # Either start of string
                                                                         |                        
                                                                         (?<unary>[+\-!])         # Capture unary operators (+, -, !) in 'unary' group
                                                                         |                       
                                                                         (?<=[^.\p{L}\p{N}_])     # Position after any char that's not a dot, letter, number, or underscore
                                                                                                  # (lookbehind - ensures we don't break existing identifiers)
+
+                                                                        (?<!\(.*\))              # Also try to start after any function call with zero to many arguments.
```
[the remark this answers is not in the document]
Comment (SchmartMaker) [reply DocxTemplater/ScriptCompiler.cs:23]: The greedyness is intentional in this case, and has been tested with nested parenthesis.

**Review by Amberg — APPROVED** [review]

**Comment by Amberg** [comment]
Yes, I'm not happy with the current solution either. Parsing with regex shouldn't really be necessary. Without the dot notation, DynamicExpresso could handle everything. That would be much more robust.

**Comment by Amberg** [comment]
> Author

Yes, I'm not happy with the current solution either. Parsing with regex shouldn't really be necessary. Without the dot notation, DynamicExpresso could handle everything. That would be much more robust.

**Comment by Amberg** [comment]
> I think I've come across an edge case with `.Func()`. I still have to confirm the case and would like to potentially continue further investigation and development in a separate PR. I would appreciate some disscussion at a later time if you had forseen the use of script syntax like `.Func()` or `..Func()`, where there's no property name after the dots.
> 
> Also, I'm considering expanding the regex to support a string array accessor, so the individual characters of a string can be accessed with the syntax `.[i]` or `.Property[i]`. I have some code stashed that addresses this, but I prefer a seperate PR for this as well.
> 
> Also, I've taken the liberty to rename function `ConditionWithStringOperators` to `ConditionWithStringOperations` since a string operator would be something like `+` and an operation would be something like `.ToUpper()` or `.Contains()`.

Yes, I'm not happy with the current solution either. Parsing with regex shouldn't really be necessary. Without the dot notation, DynamicExpresso could handle everything. That would be much more robust.
